eth/handler.go

```diff
@@ -288,22 +292,74 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		wg       sync.WaitGroup
+		done     = make(chan struct{})
+		active   int
+		quitting bool
+	)
+	// Start goroutine to track active protocol handlers.
+	wg.Add(1)
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active++
+			case <-h.handlerDoneCh:
+				active--
+			case <-done:
+				quitting = true
```

Comment:
I don't see why you are using a `done` chan, instead of just using `h.quitSync` here?

Reply:
It seemed less racy to use `done` here. The scenario I was imagining is say `h.quitSync` is closed while another routine is in the process of writing to `h.handlerStartCh`. Since there is nothing reading that channel anymore, it will sit there blocking and never close.

I wrote it in the way it is so that the select is always able to unblock any channel neccessary. Between closing `h.quitSync` and calling `peerset.close()`, the handlers _should_ be in the process of closing, even if some last minute handlers end up being registered as active.

Reply:
actually maybe that scenario isn't possible - once select chooses a branch is it guaranteed to complete uninterrupted?